sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala

```diff
@@ -941,6 +944,43 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
   }
 
   /**
+   * Throw a [[ParseException]] if the user specified incompatible SerDes through ROW FORMAT
+   * and STORED AS.
+   */
+  private def validateRowFormatFileFormat(
+      rowFormatCtx: RowFormatContext,
+      createFileFormatCtx: CreateFileFormatContext,
+      parentCtx: ParserRuleContext): Unit = {
+    if (rowFormatCtx == null || createFileFormatCtx == null) {
+      return
+    }
+    val cff = (0 until createFileFormatCtx.getChildCount)
+      .map { i => createFileFormatCtx.getChild(i).getText }
+      .mkString(" ")
+    (rowFormatCtx, createFileFormatCtx.fileFormat) match {
+      case (_, ffTable: TableFileFormatContext) =>
+        if (visitTableFileFormat(ffTable).serde.isDefined) {
+          throw operationNotAllowed(s"ROW FORMAT is not compatible with $cff", parentCtx)
+        }
+      case (rfSerde: RowFormatSerdeContext, ffGeneric: GenericFileFormatContext) =>
+        ffGeneric.identifier.getText.toLowerCase match {
+          case ("sequencefile" | "textfile" | "rcfile") => // OK
+          case _ => throw operationNotAllowed(
+            s"ROW FORMAT SERDE is not compatible with $cff", parentCtx)
+        }
+      case (rfDelimited: RowFormatDelimitedContext, ffGeneric: GenericFileFormatContext) =>
+        ffGeneric.identifier.getText.toLowerCase match {
+          case "textfile" => // OK
+          case _ => throw operationNotAllowed(
+            s"ROW FORMAT SERDE is not compatible with $cff", parentCtx)
```

Comment:
oh, sorry. I was thinking about what the error message should say.

Yea, the semantic is what we talked about.